### PR TITLE
Fixes contiki-ng/contiki-ng#774 Won't compile asm files

### PR DESCRIPTION
**Pull request by firmwareguru**
Reference issue #774.

This fix takes care of compiling both .s and .S for me.

Using `gcc` rather than `as` in `arch/cpu/arm/Makefile.arm`  to compile asm files offers all of the available preprocessor goodies, so that's what I've chosen to implement.

### CODE REVIEW

**Comment by g-oikonomou** [comment]
Hi @firmwareguru thanks for this. We want this.

Using `arm-none-eabi-gcc` instead of `-as` sounds fine to me, in fact I have done the same in a local branch too.

What I am not 100% comfortable about is blindly using `CFLAGS` as `ASFLAGS`. Could you please add some motivation? Is there a reason why not simply use `ASFLAGS`?

These are the changes that I have had to make locally:

```
arch/cpu/arm/Makefile.arm:ASFLAGS += -mthumb -mabi=aapcs -mlittle-endian
arch/cpu/arm/cortex-m/cm4/Makefile.cm4:ASFLAGS += -mcpu=cortex-m4
arch/cpu/.../Makefile.<my-new-cpu>:ASFLAGS += -mfloat-abi=hard -mfpu=fpv4-sp-d16 -specs=nano.specs
arch/cpu/.../Makefile.<my-new-cpu>:ASFLAGS += -c -x assembler-with-cpp
```

Obviously:

* Some of those are specific to the hardware I am working on and
* We are missing out on all the `-D` and `-I`,

but I just wanted to share how I am approaching the same problem.

The main line of interest here is the last one in the above snippet. This allows us to pass `-c` without touching the top-level `Makefile.include`. `-x assembler-with-cpp` allows us to put pre-processor stuff in the assembly.

Would such approach meet your requirements?

**Comment by g-oikonomou** [comment]
Also, if we are going to set `ASFLAGS` to `CFLAGS`, I'd rather do it with `+=`: Allows the user to append more.

**Comment by firmwareguru** [comment]
It seems like offering the asm files the same environment as the .c files in which to build would be the desired outcome, so that's why I used gcc rather than as, and then set ASFLAGS to CFLAGS.

But,  I realize there are possibly some more "traditional" assembly use cases that might want the file extension to determine what kind of pre-processing to apply (i.e. .s vs .S, for which assembler behavior is effectively overridden with the -x assembler-with-cpp option)

For me the "hassle" of maintaining a separate, but identical, ASFLAGS isn't warranted.

I defer to you to specify if you want to offer that extra level of control w.r.t ASFLAGS.

File: Makefile.include [hunk 1]
```diff
@@ -347,10 +349,10 @@ endif
 ifndef CUSTOM_RULE_S_TO_OBJECTDIR_O
 $(OBJECTDIR)/%.o: %.S | $(OBJECTDIR)
 	$(TRACE_AS)
-	$(Q)$(AS) $(ASFLAGS) -o $@ $<
+	$(Q)$(AS) $(ASFLAGS) -o $@ -c $<
```
Comment (g-oikonomou) [inline Makefile.include]: Please remove this change

File: Makefile.include [hunk 2]
```diff
@@ -337,6 +337,8 @@ distclean:
 
 -include $(ARCH_PATH)/platform/$(TARGET)/Makefile.customrules-$(TARGET)
 
+ASFLAGS ?= $(CFLAGS)
```
Comment (g-oikonomou) [inline Makefile.include]: This needs to change to `+=`

File: Makefile.include [hunk 3]
```diff
@@ -347,10 +349,10 @@ endif
 ifndef CUSTOM_RULE_S_TO_OBJECTDIR_O
 $(OBJECTDIR)/%.o: %.S | $(OBJECTDIR)
 	$(TRACE_AS)
-	$(Q)$(AS) $(ASFLAGS) -o $@ $<
+	$(Q)$(AS) $(ASFLAGS) -o $@ -c $<
 $(OBJECTDIR)/%.o: %.s | $(OBJECTDIR)
 	$(TRACE_AS)
-	$(Q)$(AS) $(ASFLAGS) -o $@ $<
+	$(Q)$(AS) $(ASFLAGS) -o $@ -c $<
```
Comment (g-oikonomou) [inline Makefile.include]: Please remove this change

File: Makefile.include (hunk 2 above)
Comment (g-oikonomou) [reply Makefile.include]: And also moved to `Makefile.arm`

**Comment by g-oikonomou** [comment]
@firmwareguru Hi Evan, do you think you will have time to make changes here? It'd be great to include this in the upcoming release. Please let me know if you don't have time.

**Comment by firmwareguru** [comment]
Yes, working on it!

**Comment by firmwareguru** [comment]
I'm proposing this for makefile.arm.  I don't see any use cases for not having `-c` a part of the ASFLAGS for all ARM CPUs... but you have better visibility on the big picture than I.  So tell me exactly what you want to see here and I'll do it.

```
CFLAGS += -mthumb -mabi=aapcs -mlittle-endian
CFLAGS += -Wall
CFLAGS += -std=c99
CFLAGS += -ffunction-sections -fdata-sections -fno-strict-aliasing
CFLAGS += -fshort-enums -fomit-frame-pointer -fno-builtin
ifeq ($(WERROR),1)
  CFLAGS += -Werror
endif

### Pass our CFLAGS along to assembly files in the build
ASFLAGS += $(CFLAGS)
### Specify '-c' option to assemble only and not link
ASFLAGS += -c 

LDFLAGS += -mthumb -mlittle-endian
```

**Comment by firmwareguru** [comment]
Hmm, looks like I rebased then committed another change then pushed this up, and made a mess of things.  Probably should close this and start over.

**Comment by g-oikonomou** [comment]
Just fix the branch locally and force push

**Comment by firmwareguru** [comment]
Ok, changes made and branch fixed.  Don't rebase on public branches!  

The end result is that only `makefile.arm` is modified to enable compilation of `.s` and `.S` files.

**Pull request merged**